src/main/java/com/exasol/adapter/dialects/sqlserver/SqlServerSqlGenerationVisitor.java

```diff
@@ -70,61 +59,45 @@ protected String representAsteriskInSelectList(final SqlSelectList selectList) t
     }
 
     private List<String> buildSelectStar(final SqlSelectList selectList) throws AdapterException {
-
-        if (SqlGenerationHelper.selectListRequiresCasts(selectList, this.nodeRequiresCast)) {
-            return buildSelectStarWithNodeCast(selectList);
-        } else {
-            return new ArrayList<>(Collections.singletonList("*"));
-        }
+        final Optional<List<String>> selectStartWithCastedColumns = buildSelectStarWithNodeCast(selectList);
+        return selectStartWithCastedColumns.orElseGet(() -> new ArrayList<>(Collections.singletonList("*")));
     }
 
-    private List<String> buildSelectStarWithNodeCast(final SqlSelectList selectList) throws AdapterException {
+    private Optional<List<String>> buildSelectStarWithNodeCast(final SqlSelectList selectList) throws AdapterException {
+        boolean requiresCast = false;
         final SqlStatementSelect select = (SqlStatementSelect) selectList.getParent();
         int columnId = 0;
         final List<TableMetadata> tableMetadata = new ArrayList<>();
         SqlGenerationHelper.addMetadata(select.getFromClause(), tableMetadata);
         final List<String> selectListElements = new ArrayList<>(tableMetadata.size());
         for (final TableMetadata tableMeta : tableMetadata) {
             for (final ColumnMetadata columnMeta : tableMeta.getColumns()) {
+                if (requiresCast(new SqlColumn(columnId, columnMeta))) {
+                    requiresCast = true;
```

Comment:
how about return `Optional.empty()` directly here. Then you could get rid of the non final `requiresCast`

Reply:
I can't return Optional.empty() here because I need Optional with values if the cast is required. And I need to finish the iteration till the end of the list to return it. I know this method is terrible and it's doing two things at one time. But I don't want to make more refactoring on the current branch because this branch is not about refactoring.